### PR TITLE
fix traffic graph not showing bandwidth when IPS is enabled

### PR DESCRIPTION
**Pull request by swhite2**
https://github.com/opnsense/core/issues/4272 mentions an issue with the traffic graph not showing bandwidth when IPS is enabled.

The (top half) OPNsense traffic graph relies on ifinfo for the visualization of raw packet/byte data. This information is queried by ifinfo using the sysctl net.link.generic.ifdata.[IF_INDEX].{general/linkspecific} mib. This mib in turn triggers a sysctl handler which copies generic interface information from the kernel to userspace using the if.h-supplied if_data_copy() function. Apart from copying static data, if_data_copy() also directly queries the relevant driver via the driver-supplied ifp->if_get_counter() function. 

The problem resides here. In normal (non-IPS) operational mode, drivers are responsible for delivering it's hardware statistics (such as IPACKET, OPACKETS, IBYTES, OBYTES) to any user querying via the above-mentioned sysctl. In case the driver has NOT defined counters for packets, bytes etc. (such as Intel), the system relies on a default counter, which gets incremented in iflib during the normal RX/TX routines. If IPS is enabled (thus netmap is enabled), these counters are not incremented.

This presents two options in netmap-mode (RX):
- Increment the counters as soon as the packets are passed up to the host stack. Whether they are passed up to the host stack is a decision made in userland (by the IPS).
- Increment the counters as soon as userland queries the internal netmap NIC rings. 

I have chosen the latter option in order to stay as close to the hardware as possible and show all traffic that comes in, even if a packet gets dropped by the IPS.

For the TX side, the counters are simply incremented as soon as the driver has encapped a single packet.

### CODE REVIEW

**Comment by fichtner** [comment]
Oh, nice find. Is this just an issue the way we set up IPS mode (from host to tx and from rx to host) or does the problem also exist when "bridging" interfaces?

File: sys/net/iflib.c [hunk 1]
```diff
@@ -960,6 +960,7 @@ iflib_netmap_txsync(struct netmap_kring *kring, int flags)
 	u_int const lim = kring->nkr_num_slots - 1;
 	u_int const head = kring->rhead;
 	struct if_pkt_info pi;
+	int pkt_sent, bytes_sent;
```
Comment (fichtner) [inline sys/net/iflib.c:963]: just for consistency can we make these tx_pkts and tx_bytes?

Comment (swhite2) [reply sys/net/iflib.c:963]: > Oh, nice find. Is this just an issue the way we set up IPS mode (from host to tx and from rx to host) or does the problem also exist when "bridging" interfaces?

No, IPS is correctly being setup, that is, setup as intended by the author of Netmap (e.g. igb0 -> igb0^). The actual problem resides in the TX-RX datapath. As such, you can expect this problem to persist when setting up a netmap bridge. Basically the system searches for the drivers' internal function that provides hardware statistics, if certain statistics are not present, it defaults to a default counter that is ONLY being tracked when a packet goes through the non-netmap flow. I simply added the counter in the netmap-specific TX-RX functions :-) 

The code I'm providing here prevents a driver author from having to implement counting mandatory statistics which might not even be available on the hardware.

In fact, regarding the two options I described earlier, using option 1 will cause a bridged interface to NOT report packets and bytes, while option 2 will.

> just for consistency can we make these tx_pkts and tx_bytes?

Of course! I just copied the variable names used for the counters in the non-netmap TX-RX

**Comment by fichtner** [comment]
Merged, thanks!

**Comment by toxicfirefly** [comment]
I see this issue when I enable IPS with latest release `23.1.10_1`.

**Comment by fichtner** [comment]
@toxicfirefly it's a driver specific issue

**Pull request merged**